lib/cli/lookup-files.js

```diff
@@ -24,30 +28,54 @@ const debug = require('debug')('mocha:cli:lookup-files');
  *
  * @private
  * @param {string} pathname - Pathname to check for match.
- * @return {boolean} whether pathname would be considered a hidden file.
+ * @returns {value is HiddenUnixPathname} whether pathname would be considered a hidden file.
  * @example
  * isHiddenOnUnix('.profile'); // => true
  */
-const isHiddenOnUnix = pathname => path.basename(pathname).startsWith('.');
+const isHiddenOnUnix = pathname => path.basename(pathname)[0] === '.';
+
+/** @typedef {`.${string}`} FileExtension */
 
 /**
- * Determines if pathname has a matching file extension.
+ * Normalize file extensions to ensure they have a leading period character.
  *
- * Supports multi-part extensions.
+ * @private
+ * @param {FileExtension[]|string[]|undefined|null} exts
+ * @returns {FileExtension[]}
+ */
+const normalizeFileExtensions = (exts) => {
```

Comment:
is the expected behavior for `normalizeFileExtensions(['hello.world'])` really to return `[.hello.world']`? extensions should just be the stuff after the last `.` . Maybe we are careful about this in the callers but it's something to consider.

That said, I haven't read the rest of the PR yet!